### PR TITLE
Add zero_gradient and linear_extrap boundary conditions

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Add zero_gradient and linear_extrap boundary conditions
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
All tests pass and bit-for-bit as expected.  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#0e59535ce2009a3175e67156793b3ec83f2a3c69
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [X] Yes
    - [ ] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.

Add zero_gradient and linear_extrap boundary conditions

This provides options to set the boundary conditions around the edge for a regional grid.  Only halo values on the outside halo are affected.  This is implemented largely in the HaloUpdates in ice_boundary.F90, both in the mpi and serial implementations.  The boundary conditions are computed locally for each block and don't require any explicit communications.  They are implemented consistent with mixing boundary conditions in the ew and ns directions.  The zero_gradient boundary condition sets the halo equal to the values on the boundary of the domain, basically zero gradient into the halo.  The linear_extrap boundary condition sets the halo values such that the gradient across the halo matches the gradient on the edge of the active domain, basically constant gradient.

The zero_gradient and linear_extrap boundary conditions have been implemented as an interim step toward a regional grid capability. Until restoring options are complete and the regional capability is fully tested, these boundary conditions may produce nonphysical values such as negative ice thickness.

- Added zero_gradient and linear_extrap code to HaloUpdate subroutines.
- Refactored the ice_boundary.F90 to leverage automated code generation and sync the mpi and serial versions of the code.
- Created a generate_haloUpdates.sh script to generate the various haloUpdate subroutines quickly and consistently.  This script lives in cicecore/cicedyn/infrastructure/comm/mpi and has a README.  This reduces effort and risk if the haloUpdate needs to be updated in the future.
- Synchronized the mpi and serial versions of ice_boundary.F90. Now the only difference in the files is a #define in the serial version to turn off the MPI code.
- Update the halochk subroutines to validate zero_gradient and linear_extrap boundary condition options.
- Updated some initialization to account for the new boundary condition options.
- Added explicit checks for valid boundary_type settings in ice_domain.F90.  This means checks for invalid values later in the code can be removed.
- Add set_nml.boxgauss which is a box test with a gauss distibution of ice in the center.  Added several tests leveraging this new configuration.  Added bczerogradient and bclinearextrap set_nml options and tests.
- Update the set_nml filenames for boundary condition settings, add "bc" to start of string.
- Update the test suite to cover the new boundary conditions and improve halochk unit test coverage overall.
- Update the documentation.

Improved the plabeld argument type in debug_ice so the string length is more flexible.  This issue was detected while testing.

Update initialization of uvmCD in ice_grid.F90 due to uninitialized values triggering and error in some compilers.

### CODE REVIEW

**Review by eclare108213 — CHANGES_REQUESTED** [review]
I think the 'dirichlet' boundary conditions here are actually zero-gradient neumann conditions.  Dirichlet BCs would set the boundary to a specified value or values that are a function of space but not time.  Perhaps it would be better to name these something a bit more intuitive.  E.g. a true dirichlet BC could be 'fixed' and a nonzero neumann BC could be 'gradient' or 'flux'.  The special case could be 'zero-gradient' or 'zero-flux'.  These could alternatively be specified in terms of order of the gradient extrapolation (first-order, second-order, etc).  For velocities there's the added complication of having a direction, hence 'no-slip' might be an option in that case.

File: cicecore/cicedyn/infrastructure/comm/mpi/ice_boundary.F90 [hunk 1]
```diff
@@ -850,33 +760,27 @@ function ice_HaloCreate(dist, nsBoundaryType, ewBoundaryType, &
...
-
+         call ice_HaloMsgCreate(halo, dist, iblock, -westBlock, 'north')
       endif
 
       !*** find northeast neighbor block and add to message count
 
       neBlock = ice_blocksGetNbrID(iblock, ice_blocksNorthEast,    &
                                    ewBoundaryType, nsBoundaryType)
 
+#if (1 == 0)
```
[the remark this answers is not in the document]
Comment (apcraig) [reply cicecore/cicedyn/infrastructure/comm/mpi/ice_boundary.F90]: fixed

**Comment by apcraig** [comment]
The unphysical value is happening at i=1, j=8 in a 12x12 domain.  What I think is happening is that the halo is becoming unphysical then that is influencing the first grid cell next to the halo at some point, via advection or something else.  There are no checks on the halo for valid values.

It is not easy to "test" the halo values, at least not inside the halo update.  The halo update doesn't know anything about which field it is operating on.  We could add some checks to the outer domain after the haloupdate call.  We'd have to decide which haloupdate calls to check and so forth.

What I propose we do is add some info the documentation to say these new bcs can generate unphysical values.  I think we should just consider this PR as new features in the infrastructure (that may not be totally ready to use).

Separately, I really think we're going to have to think through how we want to implement the restoring and new bcs.  I think we still have a bunch of work to do to come up with a few "options" that would mix zero gradient bcs, linear extrap bcs and restoring bcs depending on field and depending on what fields we're trying to restore.  I could see an argument made that u and v should ALWAYS be specified on the halo in open bcs with the zero gradient, never restored.  And that aice and hi can take on any of the bc options but they will have to be constrained to be >=0.  And everything else that gets a halo update cannot be restored to so has to either have zero gradient or linear extrapolation applied.  Right now, we testing ALL haloupdates with zero gradient or ALL with linear extrap or ALL with restoring.  In practice, I think none of those options are physically reasonable nor what we will ultimately offer to users to do science.

**Review by eclare108213 — APPROVED** [review]
This is awesome, thank you.  I made some suggestions for the documentation and rather than having another review round, will go ahead and approve the PR.  I also suggest copying the comment about potential nonphysical values using the new BCs into the PR merge comments.
